MediaBrowser.Providers/Music/AlbumMetadataService.cs

```diff
@@ -47,11 +47,11 @@ public AlbumMetadataService(
         protected override bool EnableUpdatingStudiosFromChildren => true;
 
         /// <inheritdoc />
-        protected override IList<BaseItem> GetChildrenForMetadataUpdates(MusicAlbum item)
+        protected override IReadOnlyList<BaseItem> GetChildrenForMetadataUpdates(MusicAlbum item)
```

Comment:
All these changes to collection types should've been in a separate PR imo.

Reply:
Those changes _are_ also made in a seperate pr but these specific are required because the new Repository classes return the values as immutable and i dont want to convert on every other place.

Most places do not modfiy those lists and only read from them so IMOP to not keep creating new lists just because, the added benefit of faster read times is worth it.

Reply:
> Those changes are also made in a seperate pr but these specific are required because the new Repository classes return the values as immutable and i dont want to convert on every other place.

Standard arrays also implement `IReadOnlyList`

> Most places do not modfiy those lists and only read from them so IMOP to not keep creating new lists just because, the added benefit of faster read times is worth it.

But what about the increase in allocations and CPU time?

BenchmarkDotNet v0.13.10, Windows 11 (10.0.22631.4317/23H2/2023Update/SunValley3)
AMD Ryzen 5 3600, 1 CPU, 12 logical and 6 physical cores
.NET SDK 8.0.300
  [Host]   : .NET 8.0.5 (8.0.524.21615), X64 RyuJIT AVX2
  .NET 8.0 : .NET 8.0.5 (8.0.524.21615), X64 RyuJIT AVX2

Job=.NET 8.0  Runtime=.NET 8.0

| Method           | Size | Mean         | Error      | StdDev     | Ratio | Gen0   | Gen1   | Allocated | Alloc Ratio |
|----------------- |----- |-------------:|-----------:|-----------:|------:|-------:|-------:|----------:|------------:|
| ToImmutableList  | 10   |    274.85 ns |   7.261 ns |  21.066 ns |  1.00 | 0.0954 |      - |     800 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToList           | 10   |     98.52 ns |   1.931 ns |   4.074 ns |  1.00 | 0.0315 |      - |     264 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToArray          | 10   |    106.46 ns |   2.037 ns |   2.264 ns |  1.00 | 0.0315 |      - |     264 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToImmutableArray | 10   |    112.05 ns |   2.263 ns |   4.466 ns |  1.00 | 0.0315 |      - |     264 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToImmutableList  | 100  |  1,706.26 ns |  34.078 ns |  79.657 ns |  1.00 | 0.7229 | 0.0057 |    6056 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToList           | 100  |    438.02 ns |   8.821 ns |  21.135 ns |  1.00 | 0.1469 |      - |    1232 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToArray          | 100  |    469.97 ns |   7.965 ns |   7.450 ns |  1.00 | 0.1431 |      - |    1200 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToImmutableArray | 100  |    459.88 ns |   7.749 ns |   7.248 ns |  1.00 | 0.1431 |      - |    1200 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToImmutableList  | 1000 | 13,716.49 ns | 271.252 ns | 516.085 ns |  1.00 | 6.7596 | 0.6409 |   56600 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToList           | 1000 |  3,532.74 ns |  59.616 ns |  58.551 ns |  1.00 | 1.0109 | 0.0153 |    8472 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToArray          | 1000 |  3,437.56 ns |  50.901 ns |  45.122 ns |  1.00 | 1.0185 |      - |    8544 B |        1.00 |
|                  |      |              |            |            |       |        |        |           |             |
| ToImmutableArray | 1000 |  3,446.18 ns |  65.145 ns |  66.899 ns |  1.00 | 1.0185 |      - |    8544 B |        1.00 |

Reply:
To be frank i really dont see us having this discussion here. I followed Trickplay and other new services which also mainly use IReadOnlyList as return values and the C# analyser is configured to prefer that interface over plain arrays on most places. Maybe we should make that a general discussion but not in this PR.